src/mode_manager.ts

```diff
@@ -3,17 +3,45 @@ import { EventEmitter } from "events";
 import { commands, Disposable } from "vscode";
 
 import { Logger } from "./logger";
-import { NeovimExtensionRequestProcessable, NeovimRedrawProcessable } from "./neovim_events_processable";
-import { findLastEvent } from "./utils";
+import { NeovimExtensionRequestProcessable } from "./neovim_events_processable";
 
 const LOG_PREFIX = "ModeManager";
 
-export class ModeManager implements Disposable, NeovimRedrawProcessable, NeovimExtensionRequestProcessable {
+// a representation of the current mode. can be read in different ways using accessors. underlying type is full name as returned by `:help mode()`
+class Mode {
+    public constructor(public full: string = "") {}
+    // mode 1-char code: n, v, V, i, s, ...
+    // converts ^v into v
+    public get char(): string {
+        return this.full.charCodeAt(0) == 22 ? "v" : this.full.charAt(0);
+    }
+    // mode full name
+    public get long(): "insert" | "visual" | "normal" {
+        switch (this.char.toLowerCase()) {
+            case "i":
+                return "insert";
+            case "v":
+                return "visual";
+            case "n":
+            default:
+                return "normal";
+        }
+    }
+    // visual mode name
+    public get visual(): "single" | "line" | "block" {
```

Comment:
the usual names are charwise, linewise, block. `:help charwise`.

```suggestion
    public get visual(): "char" | "line" | "block" {
```